core/trino-main/src/main/java/io/trino/sql/gen/columnar/OrExpressionEvaluator.java

```diff
@@ -0,0 +1,72 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.operator.project.SelectedPositions;
+import io.trino.spi.Page;
+import io.trino.sql.relational.RowExpression;
+import io.trino.sql.relational.SpecialForm;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.relational.SpecialForm.Form.OR;
+
+public final class OrExpressionEvaluator
+        implements ExpressionEvaluator
+{
+    public static Optional<Supplier<ExpressionEvaluator>> createOrExpressionEvaluator(ColumnarFilterCompiler compiler, SpecialForm specialForm)
+    {
+        checkArgument(specialForm.getForm() == OR, "specialForm %s should be OR", specialForm);
+        checkArgument(specialForm.getArguments().size() >= 2, "OR expression %s should have at least 2 arguments", specialForm);
+
+        ImmutableList.Builder<Supplier<ExpressionEvaluator>> builder = ImmutableList.builder();
+        for (RowExpression expression : specialForm.getArguments()) {
+            Optional<Supplier<ExpressionEvaluator>> subExpressionEvaluator = ExpressionEvaluator.createColumnarFilterEvaluator(expression, compiler);
+            if (subExpressionEvaluator.isEmpty()) {
+                return Optional.empty();
+            }
+            builder.add(subExpressionEvaluator.get());
+        }
+        List<Supplier<ExpressionEvaluator>> subExpressionEvaluators = builder.build();
+        return Optional.of(() -> new OrExpressionEvaluator(subExpressionEvaluators.stream().map(Supplier::get).collect(toImmutableList())));
+    }
+
+    private final List<ExpressionEvaluator> subExpressionEvaluators;
+
+    private OrExpressionEvaluator(List<ExpressionEvaluator> subExpressionEvaluators)
+    {
+        checkArgument(subExpressionEvaluators.size() >= 2, "must have at least 2 subexpressions to OR");
+        this.subExpressionEvaluators = subExpressionEvaluators;
+    }
+
+    @Override
+    public SelectedPositions evaluate(SelectedPositions activePositions, Page page)
+    {
+        SelectedPositions accumulatedPositions = null;
+        for (int index = 0; index < subExpressionEvaluators.size(); index++) {
+            ExpressionEvaluator evaluator = subExpressionEvaluators.get(index);
+            SelectedPositions selectedPositions = evaluator.evaluate(activePositions, page);
+            accumulatedPositions = accumulatedPositions == null ? selectedPositions : accumulatedPositions.union(selectedPositions);
+            if (index != subExpressionEvaluators.size() - 1) {
+                activePositions = activePositions.difference(selectedPositions);
+            }
+        }
```

Comment:
I **think** you might be able to simplify this code by splitting out the first and last  sub expressions so you don't need the two if statements.  Something like this:
```java
        SelectedPositions accumulatedPositions = subExpressionEvaluators.getFirst().evaluate(activePositions, page);
        for (int index = 1; index < subExpressionEvaluators.size() - 1; index++) {
            ExpressionEvaluator evaluator = subExpressionEvaluators.get(index);
            SelectedPositions selectedPositions = evaluator.evaluate(activePositions, page);
            accumulatedPositions = accumulatedPositions.union(selectedPositions);
            activePositions = activePositions.difference(selectedPositions);
        }
        accumulatedPositions = accumulatedPositions.union(subExpressionEvaluators.getLast().evaluate(activePositions, page);
        return accumulatedPositions;
```

Reply:
Updated with small change to include `activePositions = activePositions.difference(accumulatedPositions);` before the for loop to keep the optimization of running subsequent sub-expressions only on positions which were `false`.